### PR TITLE
deferred call submission status API refactor

### LINKED ISSUE #9243: Fusion: API improvements for handling deferred calls
During the dx testing, 2 problems were identified.
1. The `onDeferredCall` is a function. Developers tend to start using the function (i.e. `client.onDeferredCall()`) instead of define it (`client.onDeferredCall = xxx`).
2. rethrow the error doesn't look intuitive.
To overcome those, a prototype is developed in [this branch]
(https://github.com/vaadin/flow/tree/deferred-call-refactor).  Note that the boolean return value of the method indicate whether to remove the call from the queue or not.

So that the user code could look sth like
```
client.deferredCallHandler = {
  async handleDeferredCall(deferredCall: DeferredCall){
    try {
      await deferredCall.submit();  
      console.log("sucess");
    } catch (error) {
      console.log("failed");
      deferredCall.keepInTheQueue();
    }
  }
};
```

### PR DESCRIPTION
**Pull request by haijian-vaadin**
Fixes #9243. After this commit, a user is able to notify the deferred call submission status via
```
client.deferredCallSubmissionHandler = {
  async handleDeferredCallSubmission(deferredCallSubmitter: DeferredCallSubmitter){
    try{
      await deferredCallSubmitter.submit();
      console.log('submission succeeded');
    }catch(error){
      console.log('submission failed');
      deferredCallSubmitter.keepDeferredCallInTheQueue();
    }
  }
}
```
Changes:

- use a `deferredCallSubmissionHandler` class instead of a callback function, so that a user won't mistakenly calling the callback function directly instead of defining one
- instead of relying on re-throwing error to keep the deferred request in the queue, now it uses `deferredCallSubmitter.keepDeferredCallInTheQueue();`
- renamed EndpointRequest to DeferredCall
- created a common interface `EndpointCallMetaInfo` for holding the meta info of an endpoint call. Both `MiddlewareContext` and `DeferredCall` extend from it.
- split offline related code to a separate Offline.ts file

### CODE REVIEW

File: flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts [hunk 1]
```diff
@@ -182,16 +174,12 @@ export interface ConnectClientOptions {
   middlewares?: Middleware[];
 
   /**
-   * The `onDeferredCall` property value
+   * The `deferredCallSubmissionHandler` property valueThe
```
Comment (joheriks) [inline flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts:177]: Extreanous `The` at end of docstring.

File: flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts [hunk 2]
```diff
@@ -182,16 +174,12 @@ export interface ConnectClientOptions {
...
    */
-  onDeferredCall?: OnDeferredCallCallback
+  deferredCallSubmissionHandler?: DeferredCallSubmissionHandler;
```
Comment (joheriks) [inline flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts:179]: The name seems longish, how about `deferredCallHandler`?

**Review by joheriks — CHANGES_REQUESTED** [review]

File: flow-client/src/test/frontend/OfflineTests.ts [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+          fakeRequestCallFails();
+          await client.submitDeferredCalls();
+        } catch (_) {
+          // expected
+        } finally {
+          await verifyNumberOfRequsetsInTheQueue(1);
+        }
+      });
+
+      it('should remove a failed endpoint call from the queue when user catches the error without calling deferrecCall.keepInTheQueue()', async () => {
```
[the remark this answers is not in the document]
Comment (haijian-vaadin) [reply flow-client/src/test/frontend/OfflineTests.ts]: done

File: flow-client/src/test/frontend/OfflineTests.ts [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+        try {
+          await insertARequest();
+          fakeRequestCallFails();
+          await client.submitDeferredCalls();
+        } finally {
+          await verifyNumberOfRequsetsInTheQueue(0);
+        }
+      });
+
+      it('should keep a failed endpoint call in the queue when user catches the error and calls deferrecCall.keepInTheQueue()', async () => {
```
[the remark this answers is not in the document]
Comment (haijian-vaadin) [reply flow-client/src/test/frontend/OfflineTests.ts]: done

File: flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts (hunk 2 above)
Comment (haijian-vaadin) [reply flow-client/src/main/resources/META-INF/resources/frontend/Connect.ts]: renamed

**Review by joheriks — APPROVED** [review]

**Pull request merged**